{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8d4fddc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "import pandas \n",
    "pandas.set_option('mode.use_inf_as_na', True)\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "warnings.simplefilter(\"ignore\")\n",
    "\n",
    "# This is required to catch warnings when the multiprocessing module is used\n",
    "import os\n",
    "\n",
    "os.environ[\"PYTHONWARNINGS\"] = \"ignore\"\n",
    "# import pertpy as pt\n",
    "import scanpy as sc\n",
    "\n",
    "import pertpy as pt\n",
    "adata = pt.dt.kang_2018()\n",
    "from torch import tensor\n",
    "import torch, numpy as np, pandas as pd\n",
    "np.set_printoptions(linewidth=140)\n",
    "torch.set_printoptions(linewidth=140, sci_mode=False, edgeitems=7)\n",
    "pd.set_option('display.width', 140)\n",
    "import numpy as np\n",
    "import torch\n",
    "from torch.utils.data import Dataset, DataLoader\n",
    "import torch.optim as optim\n",
    "from collections import defaultdict\n",
    "rowGeneExpression2 = defaultdict(dict)\n",
    "import math\n",
    "criterion = torch.nn.CrossEntropyLoss()\n",
    "#optimizer = optim.SGD(net.parameters(), lr=0.001, momentum=0.9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f6d67306",
   "metadata": {},
   "outputs": [],
   "source": [
    "from fastai.data.transforms import RandomSplitter\n",
    "\n",
    "\n",
    "def h5adToData(adata):\n",
    "    rowGeneExpression = defaultdict(int)\n",
    "\n",
    "    sc.pp.log1p(adata)\n",
    "    sc.pp.highly_variable_genes(adata)\n",
    "#     sc.experimental.pp.recipe_pearson_residuals(adata)\n",
    "\n",
    "    sc.pp.pca(adata)\n",
    "    hv_genes = (list(adata.var[adata.var['highly_variable'] == True].index))\n",
    "    normal_genes = (list(adata.var_names))\n",
    "\n",
    "    hv_columns = [i for i,val in enumerate(normal_genes) if val in hv_genes]\n",
    "\n",
    "    df = adata.obs\n",
    "    end = 50 if True else -1\n",
    "    for column in hv_columns:\n",
    "        for row_id in range(math.floor(float(df.shape[0])))[:end]:\n",
    "            rowGeneExpression[row_id] += adata.X[row_id, column]\n",
    "\n",
    "    \n",
    "    \n",
    "    df = adata.obs\n",
    "    dependent_variables =  [rowGeneExpression[row] for row in range(df.shape[0])]\n",
    "    df['geneExpressionCount'] = dependent_variables\n",
    "    numerical_values = df.select_dtypes(include=[int, float]).values.tolist()\n",
    "    dependent_variables = [x for x in dependent_variables]\n",
    "    \n",
    "    t_dep = tensor(dependent_variables) # pertrubations\n",
    "    t_indep = tensor(numerical_values, dtype=torch.float)\n",
    "\n",
    "    independent_variables = pd.DataFrame(numerical_values)\n",
    "    vals,indices = t_indep.max(dim=0)\n",
    "    t_indep = t_indep / vals\n",
    "    #t_indep = t_indep / vals\n",
    "    \n",
    "    trn_split,val_split=RandomSplitter(seed=42)(independent_variables)\n",
    "\n",
    "    trn_indep,val_indep = t_indep[trn_split],t_indep[val_split]\n",
    "    trn_dep,val_dep = t_dep[trn_split],t_dep[val_split]\n",
    "    len(trn_indep),len(val_indep)\n",
    "    return t_dep,t_indep, trn_dep, trn_indep ,val_dep, val_indep, df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e0499927",
   "metadata": {},
   "outputs": [],
   "source": [
    "def runNN(t_dep, t_indep, trn_dep,trn_indep ,val_dep, val_indep, df):\n",
    "    indep_cols =  df.columns.tolist()\n",
    "    #dropout, svd, latent space\n",
    "    torch.manual_seed(443)\n",
    "    n_coeff = t_indep.shape[1]\n",
    "\n",
    "    def calc_preds(coeffs, indeps): return (coeffs * indeps).sum(axis=1)\n",
    "    def calc_preds(coeffs, indeps): return torch.sigmoid((indeps*coeffs).sum(axis=1))\n",
    "    #def calc_preds(coeffs, indeps): return (indeps*coeffs).sum(axis=1)\n",
    "    #def calc_preds(coeffs, indeps): return torch.sigmoid(indeps@coeffs)\n",
    "\n",
    "    def calc_loss(coeffs, indeps, deps): return torch.abs(calc_preds(coeffs, indeps)-deps).mean()\n",
    "    def init_coeffs(): return (torch.rand(n_coeff)-0.5).requires_grad_()\n",
    "    def update_coeffs(coeffs, lr): \n",
    "        coeffs.add_(-coeffs.grad * lr)\n",
    "        coeffs.grad.zero_()\n",
    "\n",
    "    def one_epoch(coeffs, lr):\n",
    "        loss = calc_loss(coeffs, t_indep, t_dep)\n",
    "        loss.backward()\n",
    "        with torch.no_grad(): update_coeffs(coeffs, lr)\n",
    "        #print(f\"{loss:.3f}\", end=\"; \")\n",
    "\n",
    "    def train_model(epochs, lr):\n",
    "        coeffs = init_coeffs()\n",
    "        for i in range(epochs):\n",
    "            one_epoch(coeffs, lr)\n",
    "        return coeffs\n",
    "\n",
    "    coeffs = train_model(33, lr=.2)\n",
    "\n",
    "    def acc(coeffs): return (val_dep.bool()==(calc_preds(coeffs, val_indep)>0.1)).float().mean()\n",
    "\n",
    "    acc(coeffs) \n",
    "    def show_coeffs(): return dict(zip(indep_cols, coeffs.requires_grad_(False)))\n",
    "    print(show_coeffs())\n",
    "    len([prob for prob in calc_preds(coeffs, t_indep) if prob > .1]) # should be 5000\n",
    "    print(len([prob for prob in calc_preds(coeffs, t_indep) if prob > .494]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77f63b02",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AdamsonWeissman2016_GSM2406675_10X001.h5ad', 'FrangiehIzar2021_RNA.h5ad', 'ReplogleWeissman2022_K562_essential.h5ad', 'ReplogleWeissman2022_rpe1.h5ad', 'TianKampmann2021_CRISPRa.h5ad', 'TianKampmann2021_CRISPRi.h5ad', 'ZhaoSims2021.h5ad']\n",
      "/home/awahab/llm-testing/data_sets/AdamsonWeissman2016_GSM2406675_10X001.h5ad\n",
      "{'perturbation': tensor(nan), 'read count': tensor(nan), 'UMI count': tensor(nan), 'tissue_type': tensor(nan), 'cell_line': tensor(nan), 'cancer': tensor(nan), 'disease': tensor(nan), 'perturbation_type': tensor(nan)}\n",
      "0\n",
      "/home/awahab/llm-testing/data_sets/FrangiehIzar2021_RNA.h5ad\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pathlib\n",
    "\n",
    "# print(pathlib.Path('yourPath.example').suffix) # '.example'\n",
    "# print(pathlib.Path(\"hello/foo.bar.tar.gz\").suffixes) # ['.bar', '.tar', '.gz']\n",
    "# print(pathlib.Path('/foo/bar.txt').stem) # 'bar'\n",
    "\n",
    "folders = '/home/awahab/llm-testing/data_sets/'\n",
    "#fp = 'FrangiehIzar2021_RNA.h5ad' - control means probably works\n",
    "fp = 'TianKampmann2021_CRISPRi.h5ad'\n",
    "fp2 = 'TianKampmann2021_CRISPRa.h5ad'\n",
    "\n",
    "adata = sc.read_h5ad(folders + fp)\n",
    "\n",
    "files = [fp for fp in os.listdir(folders) if pathlib.Path(fp).suffix == '.h5ad']\n",
    "\n",
    "\n",
    "print(files)\n",
    "def runNNWithData(file_name):\n",
    "    file_name = folders + file_name\n",
    "    adata = sc.read_h5ad(file_name)\n",
    "    print(file_name)\n",
    "    #print(adata.obs.columns)\n",
    "    t_dep, t_indep, trn_dep,val_dep,val_dep, val_indep, df = h5adToData(adata)\n",
    "    runNN(t_dep, t_indep, trn_dep,val_dep,val_dep, val_indep, df)\n",
    "    \n",
    "#runNNWithData(fp2, files)\n",
    "for f in files: runNNWithData(f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7f798940",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/awahab/llm-testing/data_sets/TianKampmann2021_CRISPRi.h5ad\n",
      "{'guide_id': tensor(-1.3326), 'perturbation': tensor(-0.1778), 'tissue_type': tensor(-0.5020), 'celltype': tensor(0.2362), 'cancer': tensor(-0.7571), 'disease': tensor(-0.0459), 'perturbation_type': tensor(-0.4703)}\n",
      "5\n"
     ]
    }
   ],
   "source": [
    "runNNWithData(files[-2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50339b66",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ac91f728",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AAACCCAAGGGTTGCA    CREBBP\n",
       "AAACCCAAGGTAGCCA    SH3RF1\n",
       "AAACCCACAAACTAGA      TAF1\n",
       "AAACCCACAGAATTCC     UBTD2\n",
       "AAACCCAGTAGCGTTT    FRMD4A\n",
       "                     ...  \n",
       "TTTGTTGCAGGTCTCG    FAM57B\n",
       "TTTGTTGCAGTTGCGC     XRCC1\n",
       "TTTGTTGGTCCCGGTA     RPL14\n",
       "TTTGTTGGTTGATGTC     PMPCA\n",
       "TTTGTTGTCTTACTGT       HTT\n",
       "Name: perturbation, Length: 32300, dtype: category\n",
       "Categories (185, object): ['ADAM10', 'ADARB2', 'AFG3L2', 'AIFM1', ..., 'ZNF292', 'ZNF592', 'ZNF599', 'control']"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad(folders + files[-2])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d8b1d94f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>guide_id</th>\n",
       "      <th>perturbation</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>celltype</th>\n",
       "      <th>cancer</th>\n",
       "      <th>disease</th>\n",
       "      <th>perturbation_type</th>\n",
       "      <th>organism</th>\n",
       "      <th>batch</th>\n",
       "      <th>nperts</th>\n",
       "      <th>ngenes</th>\n",
       "      <th>ncounts</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>percent_ribo</th>\n",
       "      <th>percent_hemo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AACAGGGTCGGAACTT</th>\n",
       "      <td>non-targeting_00186_i1</td>\n",
       "      <td>control</td>\n",
       "      <td>primary</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>False</td>\n",
       "      <td>healthy</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>human</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>4558</td>\n",
       "      <td>13654.0</td>\n",
       "      <td>7.228651</td>\n",
       "      <td>8.664128</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACCTGAGTTAGGCTT</th>\n",
       "      <td>non-targeting_00110_i1</td>\n",
       "      <td>control</td>\n",
       "      <td>primary</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>False</td>\n",
       "      <td>healthy</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>human</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5913</td>\n",
       "      <td>24923.0</td>\n",
       "      <td>4.489829</td>\n",
       "      <td>12.783373</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACCTTTAGGTTCCAT</th>\n",
       "      <td>non-targeting_00110_i1</td>\n",
       "      <td>control</td>\n",
       "      <td>primary</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>False</td>\n",
       "      <td>healthy</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>human</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>4272</td>\n",
       "      <td>11310.0</td>\n",
       "      <td>5.685234</td>\n",
       "      <td>6.277630</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACGTCAAGTTCTCTT</th>\n",
       "      <td>non-targeting_00574_i1</td>\n",
       "      <td>control</td>\n",
       "      <td>primary</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>False</td>\n",
       "      <td>healthy</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>human</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>4982</td>\n",
       "      <td>16091.0</td>\n",
       "      <td>6.332733</td>\n",
       "      <td>7.078491</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AATTTCCTCGCCTAGG</th>\n",
       "      <td>non-targeting_00692_i1</td>\n",
       "      <td>control</td>\n",
       "      <td>primary</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>False</td>\n",
       "      <td>healthy</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>human</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>4849</td>\n",
       "      <td>15430.0</td>\n",
       "      <td>4.141283</td>\n",
       "      <td>7.738172</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGGTTTCAGGACTAG</th>\n",
       "      <td>non-targeting_00655_i1</td>\n",
       "      <td>control</td>\n",
       "      <td>primary</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>False</td>\n",
       "      <td>healthy</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>human</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>5319</td>\n",
       "      <td>18188.0</td>\n",
       "      <td>7.037607</td>\n",
       "      <td>9.528260</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTTCAGTGTAAACA</th>\n",
       "      <td>non-targeting_00574_i1</td>\n",
       "      <td>control</td>\n",
       "      <td>primary</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>False</td>\n",
       "      <td>healthy</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>human</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4065</td>\n",
       "      <td>10626.0</td>\n",
       "      <td>6.446452</td>\n",
       "      <td>6.070017</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTACGTCACTGTCGG</th>\n",
       "      <td>non-targeting_00655_i1</td>\n",
       "      <td>control</td>\n",
       "      <td>primary</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>False</td>\n",
       "      <td>healthy</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>human</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>3974</td>\n",
       "      <td>10514.0</td>\n",
       "      <td>4.441697</td>\n",
       "      <td>6.572189</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCAGTTCCTAGCTC</th>\n",
       "      <td>non-targeting_00883_i1</td>\n",
       "      <td>control</td>\n",
       "      <td>primary</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>False</td>\n",
       "      <td>healthy</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>human</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>5364</td>\n",
       "      <td>18802.0</td>\n",
       "      <td>4.286778</td>\n",
       "      <td>8.866078</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGATCCACGAAAGC</th>\n",
       "      <td>non-targeting_00110_i1</td>\n",
       "      <td>control</td>\n",
       "      <td>primary</td>\n",
       "      <td>iPSC-induced neuron</td>\n",
       "      <td>False</td>\n",
       "      <td>healthy</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>human</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4295</td>\n",
       "      <td>12525.0</td>\n",
       "      <td>7.321357</td>\n",
       "      <td>5.996008</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>437 rows ￗ 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                guide_id perturbation tissue_type             celltype  cancer  disease perturbation_type organism batch  \\\n",
       "AACAGGGTCGGAACTT  non-targeting_00186_i1      control     primary  iPSC-induced neuron   False  healthy            CRISPR    human     1   \n",
       "AACCTGAGTTAGGCTT  non-targeting_00110_i1      control     primary  iPSC-induced neuron   False  healthy            CRISPR    human     1   \n",
       "AACCTTTAGGTTCCAT  non-targeting_00110_i1      control     primary  iPSC-induced neuron   False  healthy            CRISPR    human     1   \n",
       "AACGTCAAGTTCTCTT  non-targeting_00574_i1      control     primary  iPSC-induced neuron   False  healthy            CRISPR    human     1   \n",
       "AATTTCCTCGCCTAGG  non-targeting_00692_i1      control     primary  iPSC-induced neuron   False  healthy            CRISPR    human     1   \n",
       "...                                  ...          ...         ...                  ...     ...      ...               ...      ...   ...   \n",
       "TTGGTTTCAGGACTAG  non-targeting_00655_i1      control     primary  iPSC-induced neuron   False  healthy            CRISPR    human     4   \n",
       "TTGTTCAGTGTAAACA  non-targeting_00574_i1      control     primary  iPSC-induced neuron   False  healthy            CRISPR    human     4   \n",
       "TTTACGTCACTGTCGG  non-targeting_00655_i1      control     primary  iPSC-induced neuron   False  healthy            CRISPR    human     4   \n",
       "TTTCAGTTCCTAGCTC  non-targeting_00883_i1      control     primary  iPSC-induced neuron   False  healthy            CRISPR    human     4   \n",
       "TTTGATCCACGAAAGC  non-targeting_00110_i1      control     primary  iPSC-induced neuron   False  healthy            CRISPR    human     4   \n",
       "\n",
       "                  nperts  ngenes  ncounts  percent_mito  percent_ribo  percent_hemo  \n",
       "AACAGGGTCGGAACTT       0    4558  13654.0      7.228651      8.664128           0.0  \n",
       "AACCTGAGTTAGGCTT       0    5913  24923.0      4.489829     12.783373           0.0  \n",
       "AACCTTTAGGTTCCAT       0    4272  11310.0      5.685234      6.277630           0.0  \n",
       "AACGTCAAGTTCTCTT       0    4982  16091.0      6.332733      7.078491           0.0  \n",
       "AATTTCCTCGCCTAGG       0    4849  15430.0      4.141283      7.738172           0.0  \n",
       "...                  ...     ...      ...           ...           ...           ...  \n",
       "TTGGTTTCAGGACTAG       0    5319  18188.0      7.037607      9.528260           0.0  \n",
       "TTGTTCAGTGTAAACA       0    4065  10626.0      6.446452      6.070017           0.0  \n",
       "TTTACGTCACTGTCGG       0    3974  10514.0      4.441697      6.572189           0.0  \n",
       "TTTCAGTTCCTAGCTC       0    5364  18802.0      4.286778      8.866078           0.0  \n",
       "TTTGATCCACGAAAGC       0    4295  12525.0      7.321357      5.996008           0.0  \n",
       "\n",
       "[437 rows x 15 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[adata.obs['perturbation'] == 'control' or adata.obs['perturbation'] == 'ctrl ' or adata.obs['perturbation'] == '*']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75763ccc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# model = scgen.SCGEN(adata_t, n_hidden=800, n_latent=100, n_layers=2)\n",
    "# mdata = pt.dt.papalexi_2021()\n",
    "# for col in mdata.obs: print(col)\n",
    "# sc.pp.normalize_total(mdata[\"rna\"])\n",
    "# sc.pp.log1p(mdata[\"rna\"])\n",
    "# sc.pp.highly_variable_genes(mdata[\"rna\"], subset=True)\n",
    "# mu.prot.pp.clr(mdata[\"adt\"])\n",
    "# sc.pp.pca(mdata[\"rna\"])\n",
    "# We calculate neighbors with the cosine distance similarly to the original Seurat implementation\n",
    "# sc.pp.neighbors(mdata[\"rna\"], metric=\"cosine\")\n",
    "# sc.tl.umap(mdata[\"rna\"])\n",
    "# sc.pl.umap(mdata[\"rna\"], color=[\"replicate\", \"Phase\", \"perturbation\"])\n",
    "# ms = pt.tl.Mixscape()\n",
    "\n",
    "# ms.perturbation_signature(\n",
    "#     mdata[\"rna\"],\n",
    "#     pert_key=\"perturbation\",\n",
    "#     control=\"NT\",\n",
    "#     split_by=\"replicate\",\n",
    "#     n_neighbors=20,\n",
    "# )\n",
    "# ms.mixscape(adata=mdata[\"rna\"], control=\"NT\", labels=\"gene_target\", layer=\"X_pert\")\n",
    "# mdata[\"rna\"].obs\n",
    "# pt.pl.ms.perturbscore(\n",
    "#     adata=mdata[\"rna\"], labels=\"gene_target\", target_gene=\"IFNGR2\", color=\"orange\"\n",
    "# )\n",
    "\n",
    "#adata.var\n",
    "#preprocess\n",
    "#batchcontrol\n",
    "#clustering\n",
    "#cell type annotation\n",
    "#embedding\n",
    "#reverse transcriptome\n",
    "#differential expression\n",
    "# pt.pl.ms.heatmap(\n",
    "#     adata=mdata[\"rna\"],\n",
    "#     labels=\"gene_target\",\n",
    "#     target_gene=\"IFNGR2\",\n",
    "#     layer=\"X_pert\",\n",
    "#     control=\"NT\",\n",
    "# )\n",
    "import scgen\n",
    "adata.obs.rename({\"label\": \"condition\"}, axis=1, inplace=True)\n",
    "adata.obs[\"condition\"].replace({\"ctrl\": \"control\", \"stim\": \"stimulated\"}, inplace=True)\n",
    "\n",
    "#  are very local features enough or do we need global context? How much variation is there and what form does it take? What variation is spurious and could be preprocessed out? Does spatial position matter or do we want to average pool it out? How much does detail matter and how far could we afford to downsample the images? How noisy are the labels?\n",
    "\n",
    "\n",
    "# https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02605-1\n",
    "\n",
    "# adata_t = adata[\n",
    "#     ~(\n",
    "#         (adata.obs[\"cell_type\"] == \"CD4 T cells\")\n",
    "#         & (adata.obs[\"condition\"] == \"stimulated\")\n",
    "#     )\n",
    "# ].copy()\n",
    "\n",
    "# cd4t_stim = adata[\n",
    "#     (\n",
    "#         (adata.obs[\"cell_type\"] == \"CD4 T cells\")\n",
    "#         & (adata.obs[\"condition\"] == \"stimulated\")\n",
    "#     )\n",
    "# ].copy()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# pred.obs\n",
    "\n",
    "# train = adata\n",
    "# ctrl_adata = adata[\n",
    "#     ((adata.obs[\"cell_type\"] == \"CD4 T cells\") & (adata.obs[\"condition\"] == \"control\"))\n",
    "# ]\n",
    "# stim_adata = train[((train.obs['cell_type'] == 'CD4T') & (train.obs['condition'] == 'stimulated'))]\n",
    "\n",
    "# # concatenate pred, control and real CD4 T cells in to one object\n",
    "# eval_adata = ctrl_adata.concatenate(cd4t_stim, pred)\n",
    "# sc.tl.pca(eval_adata)\n",
    "# sc.pl.pca(eval_adata, color=\"condition\", frameon=False)\n",
    "# cd4t_adata = adata[adata.obs[\"cell_type\"] == \"CD4 T cells\"]\n",
    "# sc.tl.rank_genes_groups(cd4t_adata, groupby=\"condition\", method=\"wilcoxon\")\n",
    "# diff_genes = cd4t_adata.uns[\"rank_genes_groups\"][\"names\"][\"stimulated\"]\n",
    "# diff_genes\n",
    "# pred, delta = model.predict(\n",
    "#     ctrl_key=\"control\", stim_key=\"stimulated\", celltype_to_predict=\"CD4 T cells\"\n",
    "# )\n",
    "\n",
    "# # we annotate the predicted cells to distinguish them later from ground truth cells.\n",
    "# pred.obs[\"condition\"] = \"prediction\"\n",
    "# print(pred)\n",
    "# sc.pp.neighbors(adata_t, use_rep=\"scgen\")\n",
    "# sc.tl.umap(adata)\n",
    "# adata_t.obsm[\"scgen\"] = model.get_latent_representation()\n",
    "# model.train(\n",
    "#     max_epochs=1, batch_size=32, early_stopping=True, early_stopping_patience=25\n",
    "# )\n",
    "\n",
    "\n",
    "# scgen.SCGEN.setup_anndata(adata_t, batch_key=\"condition\", labels_key=\"cell_type\")\n",
    "\n",
    "\n",
    "\n",
    "# mdata[\"adt\"].obs[\"mixscape_class_global\"] = mdata[\"rna\"].obs[\"mixscape_class_global\"]\n",
    "# pt.pl.ms.violin(\n",
    "#     adata=mdata[\"adt\"],\n",
    "#     target_gene_idents=[\"NT\", \"JAK2\", \"STAT1\", \"IFNGR1\", \"IFNGR2\", \"IRF1\"],\n",
    "#     keys=\"PDL1\",\n",
    "#     groupby=\"gene_target\",\n",
    "#     hue=\"mixscape_class_global\",\n",
    "# )\n",
    "# sc.settings.set_figure_params(figsize=(10, 10))\n",
    "# pt.pl.ms.violin(\n",
    "#     adata=mdata[\"rna\"],\n",
    "#     target_gene_idents=[\"NT\", \"IFNGR2 NP\", \"IFNGR2 KO\"],\n",
    "#     groupby=\"mixscape_class\",\n",
    "# )\n",
    "#rna velocity -> predicts the future state \n",
    "#predict the probability of unseen perturbations\n",
    "#test: first data set => output list of perturbations \n",
    "# take a 2nd dataset and get list of perturbations + their probability \n",
    "# ms.lda(adata=mdata[\"rna\"], labels=\"gene_target\", layer=\"X_pert\")\n",
    "\n",
    "#representation learning (in particular, self-supervised, multi-view, and transfer learning\n",
    "#https://registry.opendata.aws/tabula-muris/#usageexamples\n",
    "\n",
    "#see what preprocessing is needed to get better accuracy\n",
    "\n",
    "##load data\n",
    "\n",
    "##import premade model => output list of predicted unseen perturbation\n",
    "#sc.pp.calculate_qc_metrics(adata)\n",
    "\n",
    "#write custom model. get better accuracy\n",
    "\n",
    "#https://www.10xgenomics.com/resources/datasets/10-k-a-375-cells-transduced-with-1-non-target-and-1-target-sg-rna-dual-indexed-3-1-standard-4-0-0\n",
    "# !pip install --quiet hyperopt\n",
    "# !pip install --quiet \"ray[tune]\"\n",
    "# !pip install --quiet scvi-colab\n",
    "# from scvi_colab import install\n",
    "\n",
    "# install()\n",
    "\n",
    "\n",
    "# import ray\n",
    "# import scanpy as sc\n",
    "# import scvi\n",
    "# from ray import tune\n",
    "# from scvi import autotune\n",
    "# model_cls = scvi.model.SCVI\n",
    "# model_cls.setup_anndata(adata)\n",
    "# scvi_tuner = autotune.ModelTuner(model_cls)\n",
    "\n",
    "\n",
    "# # pt.pl.ms.lda(adata=mdata[\"rna\"])\n",
    "# #https://zenodo.org/record/7058382\n",
    "# \n",
    "# fp1 = 'SC3_v3_NextGem_DI_CellPlex_CRISPR_A549_30K_Multiplex_count_raw_molecule_info.h5'\n",
    "# fp2 = 'SC3_v3_NextGem_DI_CRISPR_A549_5K_Multiplex_count_raw_molecule_info.h5'\n",
    "\n",
    "\n",
    "# import h5py\n",
    "# import anndata\n",
    "\n",
    "# # Read the .h5 File\n",
    "# def explore_h5py_group(group, indent=0):\n",
    "#     \"\"\"Recursively print the contents of an h5py group/dataset.\"\"\"\n",
    "#     #print(group.data.name)\n",
    "#     items = sorted(group.items())\n",
    "#     for name, item in items:\n",
    "#         if isinstance(item, h5py.Dataset):  # Check if item is a dataset\n",
    "#             print(\"  \" * indent + f\"Dataset: {name} (Shape: {item.shape}, Dtype: {item.dtype})\")\n",
    "#         elif isinstance(item, h5py.Group):  # Check if item is a group\n",
    "#             print(\"  \" * indent + f\"Group: {name}\")\n",
    "#             explore_h5py_group(item, indent + 1)  # Recursive call to explore subgroups\n",
    "\n",
    "# # Open your HDF5 file\n",
    "# with h5py.File(folders + fp2, 'r') as f:\n",
    "#     explore_h5py_group(f)\n",
    "    \n",
    "# #hv_genes = filtered_keys = df[df['column_name'] == True].index.tolist()\n",
    "\n",
    "# #normal_genes = filtered_keys = df[df['column_name'] == False].index.tolist()\n",
    "# #https://karpathy.github.io/2019/04/25/recipe/\n",
    "# #https://karpathy.medium.com/yes-you-should-understand-backprop-e2f06eab496b\n",
    "# #check bookmarks\n",
    "# #https://www.youtube.com/watch?v=qlLChbHhbg4&ab_channel=AndrejKarpathy\n",
    "# #print(hv_genes, normal_genes)\n",
    "# #https://github.com/karpathy/nn-zero-to-hero\n",
    "\n",
    "# # import numpy as np\n",
    "# # import matplotlib.pyplot as plt\n",
    "# # %matplotlib inline\n",
    "\n",
    "\n",
    "# # plt.hist(hv_columns, bins=30, edgecolor=\"k\")  # `bins` determines the number of bins/bars.\n",
    "# # plt.title(\"Histogram of Data\")\n",
    "# # plt.xlabel(\"Value\")\n",
    "# # plt.ylabel(\"Frequency\")\n",
    "# # plt.show()\n",
    "# http://projects.sanderlab.org/pertbio/\n",
    "#https://nbisweden.github.io/workshop-scRNAseq/\n",
    "\n",
    "# We create a copy of the object to recalculate the PCA.\n",
    "# Alternatively we could replace the X of the RNA part of our MuData object with the `X_pert` layer.\n",
    "# adata_pert = mdata[\"rna\"].copy()\n",
    "# adata_pert.X = adata_pert.layers[\"X_pert\"]\n",
    "# sc.pp.pca(adata_pert)\n",
    "# sc.pp.neighbors(adata_pert, metric=\"cosine\")\n",
    "# sc.tl.umap(adata_pert)\n",
    "# sc.pl.umap(adata_pert, color=[\"replicate\", \"Phase\", \"perturbation\"])\n",
    "\n",
    "# r2_value = model.reg_mean_plot(\n",
    "#     eval_adata,\n",
    "#     axis_keys={\"x\": \"predicted stimulated\", \"y\": \"stimulated\"},\n",
    "#     gene_list=diff_genes[:10],\n",
    "#     top_100_genes=diff_genes,\n",
    "#     labels={\"x\": \"predicted\", \"y\": \"ground truth\"},\n",
    "#     show=True,\n",
    "#     legend=False,\n",
    "# )\n",
    "# import pertpy as pt\n",
    "# import muon as mu\n",
    "# import scanpy as sc\n",
    "# sc.pl.violin(eval_adata, keys=\"ISG15\", groupby=\"condition\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "21b7957a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "downloading:  https://zenodo.org/record/7416068/files/GasperiniShendure2019_highMOI.h5ad?download=1\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 58\u001b[0m\n\u001b[1;32m     55\u001b[0m       \u001b[38;5;28;01mfor\u001b[39;00m data \u001b[38;5;129;01min\u001b[39;00m r:\n\u001b[1;32m     56\u001b[0m         f\u001b[38;5;241m.\u001b[39mwrite(data)\n\u001b[0;32m---> 58\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m link \u001b[38;5;129;01min\u001b[39;00m url: download_url(link)\n\u001b[1;32m     59\u001b[0m     \u001b[38;5;66;03m#https://zenodo.org/record/7058382\u001b[39;00m\n\u001b[1;32m     60\u001b[0m     \n\u001b[1;32m     61\u001b[0m \u001b[38;5;66;03m#find data set with some perturbed and some not\u001b[39;00m\n\u001b[1;32m     62\u001b[0m \u001b[38;5;66;03m#train on a half with some of both \u001b[39;00m\n\u001b[1;32m     63\u001b[0m \u001b[38;5;66;03m#apply to other half and see if predictions are true\u001b[39;00m\n\u001b[1;32m     66\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mrequests\u001b[39;00m\n",
      "Cell \u001b[0;32mIn[9], line 55\u001b[0m, in \u001b[0;36mdownload_url\u001b[0;34m(url)\u001b[0m\n\u001b[1;32m     53\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m r\u001b[38;5;241m.\u001b[39mstatus_code \u001b[38;5;241m==\u001b[39m requests\u001b[38;5;241m.\u001b[39mcodes\u001b[38;5;241m.\u001b[39mok:\n\u001b[1;32m     54\u001b[0m   \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(file_name, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwb\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m f:\n\u001b[0;32m---> 55\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m data \u001b[38;5;129;01min\u001b[39;00m r:\n\u001b[1;32m     56\u001b[0m       f\u001b[38;5;241m.\u001b[39mwrite(data)\n",
      "File \u001b[0;32m~/anaconda3/envs/shit/lib/python3.10/site-packages/requests/models.py:816\u001b[0m, in \u001b[0;36mResponse.iter_content.<locals>.generate\u001b[0;34m()\u001b[0m\n\u001b[1;32m    814\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mraw, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstream\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m    815\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 816\u001b[0m         \u001b[38;5;28;01myield from\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mraw\u001b[38;5;241m.\u001b[39mstream(chunk_size, decode_content\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m    817\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m ProtocolError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    818\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m ChunkedEncodingError(e)\n",
      "File \u001b[0;32m~/anaconda3/envs/shit/lib/python3.10/site-packages/urllib3/response.py:628\u001b[0m, in \u001b[0;36mHTTPResponse.stream\u001b[0;34m(self, amt, decode_content)\u001b[0m\n\u001b[1;32m    626\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    627\u001b[0m     \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_fp_closed(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fp):\n\u001b[0;32m--> 628\u001b[0m         data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mamt\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mamt\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdecode_content\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    630\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m data:\n\u001b[1;32m    631\u001b[0m             \u001b[38;5;28;01myield\u001b[39;00m data\n",
      "File \u001b[0;32m~/anaconda3/envs/shit/lib/python3.10/site-packages/urllib3/response.py:567\u001b[0m, in \u001b[0;36mHTTPResponse.read\u001b[0;34m(self, amt, decode_content, cache_content)\u001b[0m\n\u001b[1;32m    564\u001b[0m fp_closed \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fp, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mclosed\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m    566\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_error_catcher():\n\u001b[0;32m--> 567\u001b[0m     data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fp_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mamt\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m fp_closed \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    568\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m amt \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    569\u001b[0m         flush_decoder \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/shit/lib/python3.10/site-packages/urllib3/response.py:533\u001b[0m, in \u001b[0;36mHTTPResponse._fp_read\u001b[0;34m(self, amt)\u001b[0m\n\u001b[1;32m    530\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m buffer\u001b[38;5;241m.\u001b[39mgetvalue()\n\u001b[1;32m    531\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    532\u001b[0m     \u001b[38;5;66;03m# StringIO doesn't like amt=None\u001b[39;00m\n\u001b[0;32m--> 533\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mamt\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mif\u001b[39;00m amt \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fp\u001b[38;5;241m.\u001b[39mread()\n",
      "File \u001b[0;32m~/anaconda3/envs/shit/lib/python3.10/http/client.py:466\u001b[0m, in \u001b[0;36mHTTPResponse.read\u001b[0;34m(self, amt)\u001b[0m\n\u001b[1;32m    463\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlength \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m amt \u001b[38;5;241m>\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlength:\n\u001b[1;32m    464\u001b[0m     \u001b[38;5;66;03m# clip the read to the \"end of response\"\u001b[39;00m\n\u001b[1;32m    465\u001b[0m     amt \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlength\n\u001b[0;32m--> 466\u001b[0m s \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mamt\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    467\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m s \u001b[38;5;129;01mand\u001b[39;00m amt:\n\u001b[1;32m    468\u001b[0m     \u001b[38;5;66;03m# Ideally, we would raise IncompleteRead if the content-length\u001b[39;00m\n\u001b[1;32m    469\u001b[0m     \u001b[38;5;66;03m# wasn't satisfied, but it might break compatibility.\u001b[39;00m\n\u001b[1;32m    470\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_close_conn()\n",
      "File \u001b[0;32m~/anaconda3/envs/shit/lib/python3.10/socket.py:705\u001b[0m, in \u001b[0;36mSocketIO.readinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    703\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[1;32m    704\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 705\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sock\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrecv_into\u001b[49m\u001b[43m(\u001b[49m\u001b[43mb\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    706\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m timeout:\n\u001b[1;32m    707\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_timeout_occurred \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/shit/lib/python3.10/ssl.py:1274\u001b[0m, in \u001b[0;36mSSLSocket.recv_into\u001b[0;34m(self, buffer, nbytes, flags)\u001b[0m\n\u001b[1;32m   1270\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m flags \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m   1271\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   1272\u001b[0m           \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnon-zero flags not allowed in calls to recv_into() on \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m\n\u001b[1;32m   1273\u001b[0m           \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m)\n\u001b[0;32m-> 1274\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnbytes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbuffer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1275\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1276\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39mrecv_into(buffer, nbytes, flags)\n",
      "File \u001b[0;32m~/anaconda3/envs/shit/lib/python3.10/ssl.py:1130\u001b[0m, in \u001b[0;36mSSLSocket.read\u001b[0;34m(self, len, buffer)\u001b[0m\n\u001b[1;32m   1128\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1129\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m buffer \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 1130\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sslobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbuffer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1131\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1132\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sslobj\u001b[38;5;241m.\u001b[39mread(\u001b[38;5;28mlen\u001b[39m)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "url = [\n",
    "#     \"https://zenodo.org/record/7416068/files/AdamsonWeissman2016_GSM2406675_10X001.h5ad?download=1\",\n",
    "#     \"https://zenodo.org/record/7416068/files/AdamsonWeissman2016_GSM2406677_10X005.h5ad?download=1\",\n",
    "#     \"https://zenodo.org/record/7416068/files/AdamsonWeissman2016_GSM2406681_10X010.h5ad?download=1\",\n",
    "#     \"https://zenodo.org/record/7416068/files/AissaBenevolenskaya2021.h5ad?download=1\",\n",
    "#     \"https://zenodo.org/record/7416068/files/ChangYe2021.h5ad?download=1\",\n",
    "#     \"https://zenodo.org/record/7416068/files/DatlingerBock2017.h5ad?download=1\",\n",
    "#     \"https://zenodo.org/record/7416068/files/DatlingerBock2021.h5ad?download=1\",\n",
    "#     \"https://zenodo.org/record/7416068/files/DixitRegev2016.h5ad?download=1\",\n",
    "#     \"https://zenodo.org/record/7416068/files/FrangiehIzar2021_protein.h5ad?download=1\",\n",
    "#     \"https://zenodo.org/record/7416068/files/FrangiehIzar2021_RNA.h5ad?download=1\",\n",
    "#     \"https://zenodo.org/record/7416068/files/GasperiniShendure2019_atscale.h5ad?download=1\",\n",
    "    \n",
    "    \"https://zenodo.org/record/7416068/files/GasperiniShendure2019_highMOI.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/GasperiniShendure2019_lowMOI.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/GehringPachter2019.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/McFarlandTsherniak2020.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/NormanWeissman2019_filtered.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/PapalexiSatija2021_eccite_arrayed_protein.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/PapalexiSatija2021_eccite_arrayed_RNA.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/PapalexiSatija2021_eccite_protein.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/PapalexiSatija2021_eccite_RNA.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/ReplogleWeissman2022_K562_essential.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/ReplogleWeissman2022_K562_gwps.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/ReplogleWeissman2022_rpe1.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/SchiebingerLander2019_GSE106340.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/SchiebingerLander2019_GSE115943.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/SchraivogelSteinmetz2020_TAP_SCREEN__chromosome_11_screen.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/SchraivogelSteinmetz2020_TAP_SCREEN__chromosome_8_screen.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/ShifrutMarson2018.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/SrivatsanTrapnell2020_sciplex2.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/SrivatsanTrapnell2020_sciplex3.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/SrivatsanTrapnell2020_sciplex4.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/TianKampmann2019_day7neuron.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/TianKampmann2019_iPSC.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/TianKampmann2021_CRISPRa.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/TianKampmann2021_CRISPRi.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/WeinrebKlein2020.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/XieHon2017.h5ad?download=1\",\n",
    "    \"https://zenodo.org/record/7416068/files/ZhaoSims2021.h5ad?download=1\"\n",
    "]\n",
    "\n",
    "import requests\n",
    " \n",
    "def download_url(url):\n",
    "  print(\"downloading: \",url)\n",
    "  # assumes that the last segment after the / represents the file name\n",
    "  # if url is abc/xyz/file.txt, the file name will be file.txt\n",
    "  file_name_start_pos = url.rfind(\"/\") + 1\n",
    "  file_name = url[file_name_start_pos:]\n",
    " \n",
    "  r = requests.get(url, stream=True)\n",
    "  if r.status_code == requests.codes.ok:\n",
    "    with open(file_name, 'wb') as f:\n",
    "      for data in r:\n",
    "        f.write(data)\n",
    "\n",
    "for link in url: download_url(link)\n",
    "    #https://zenodo.org/record/7058382\n",
    "    \n",
    "#find data set with some perturbed and some not\n",
    "#train on a half with some of both \n",
    "#apply to other half and see if predictions are true\n",
    "\n",
    "\n",
    "import requests\n",
    "from multiprocessing.pool import ThreadPool\n",
    "\n",
    "def download_url(url):\n",
    "  print(\"downloading: \",url)\n",
    "  # assumes that the last segment after the / represents the file name\n",
    "  # if url is abc/xyz/file.txt, the file name will be file.txt\n",
    "  file_name_start_pos = url.rfind(\"/\") + 1\n",
    "  file_name = url[file_name_start_pos:]\n",
    "\n",
    "  r = requests.get(url, stream=True)\n",
    "  if r.status_code == requests.codes.ok:\n",
    "    with open(file_name, 'wb') as f:\n",
    "      for data in r:\n",
    "        f.write(data)\n",
    "  return url\n",
    "\n",
    "# results = ThreadPool(5).imap_unordered(download_url, url)\n",
    "# for r in results:\n",
    "#     print(r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4cd2dbdf",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'url' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 9\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01maiohttp\u001b[39;00m  \u001b[38;5;66;03m# pip install aiohttp\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01maiofile\u001b[39;00m  \u001b[38;5;66;03m# pip install aiofile\u001b[39;00m\n\u001b[0;32m----> 9\u001b[0m FILES_PATH \u001b[38;5;241m=\u001b[39m \u001b[43murl\u001b[49m\n\u001b[1;32m     12\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mdownload_files_from_report\u001b[39m(urls):\n\u001b[1;32m     13\u001b[0m     sema \u001b[38;5;241m=\u001b[39m asyncio\u001b[38;5;241m.\u001b[39mBoundedSemaphore(\u001b[38;5;241m5\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'url' is not defined"
     ]
    }
   ],
   "source": [
    "# ! pip install aiohttp\n",
    "# ! pip install aiofile\n",
    "import os\n",
    "import asyncio\n",
    "import aiohttp  # pip install aiohttp\n",
    "import aiofile  # pip install aiofile\n",
    "\n",
    "\n",
    "FILES_PATH = url\n",
    "\n",
    "\n",
    "def download_files_from_report(urls):\n",
    "    sema = asyncio.BoundedSemaphore(5)\n",
    "\n",
    "    async def fetch_file(session, url):\n",
    "        fname = url.split(\"/\")[-1]\n",
    "        async with sema:\n",
    "            async with session.get(url) as resp:\n",
    "                assert resp.status == 200\n",
    "                data = await resp.read()\n",
    "\n",
    "        async with aiofile.async_open(\n",
    "            os.path.join(FILES_PATH, fname), \"wb\"\n",
    "        ) as outfile:\n",
    "            await outfile.write(data)\n",
    "\n",
    "    async def main():\n",
    "        async with aiohttp.ClientSession() as session:\n",
    "            tasks = [fetch_file(session, url) for url in urls]\n",
    "            await asyncio.gather(*tasks)\n",
    "\n",
    "    loop = asyncio.get_event_loop()\n",
    "    loop.run_until_complete(main())\n",
    "    loop.close()\n",
    "for fp in FILES_PATH: download_files_from_report(fp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f571fb1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#https://zenodo.org/record/7058382\n",
    "#http://projects.sanderlab.org/scperturb/datavzrd/scPerturb_vzrd_v1/dataset_info/index_1.html\n",
    "#https://nbisweden.github.io/workshop-scRNAseq/"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
